Source/Core/Core/IOS/FS/FileSystemProxy.cpp

```diff
@@ -0,0 +1,592 @@
+
+  m_fd_map[request.fd] = {request.gid, request.uid, backend_fd->Release()};
+  std::strncpy(m_fd_map[request.fd].name.data(), request.path.c_str(), 64);
+  return GetFSReply(IPC_SUCCESS, ticks);
+}
+
+IPCCommandResult FS::Close(u32 fd)
+{
+  u64 ticks = 0;
+  if (m_fd_map[fd].fs_fd != INVALID_FD)
+  {
+    if (fd == m_cache_fd)
+    {
+      ticks += SimulateFlushFileCache();
+      m_cache_fd = INVALID_FD;
+    }
+
+    if (m_fd_map[fd].superblock_flush_needed)
+      ticks += SUPERBLOCK_WRITE_TICKS;
+
+    const ResultCode result = m_ios.GetFS()->Close(m_fd_map[fd].fs_fd);
+    LogResult(StringFromFormat("Close(%s)", m_fd_map[fd].name.data()), result);
+    m_fd_map.erase(fd);
+    if (result != ResultCode::Success)
+      return GetFSReply(ConvertResult(result));
+  }
+
+  m_fd_map.erase(fd);
+  return GetFSReply(IPC_SUCCESS, ticks);
+}
+
+u64 FS::SimulatePopulateFileCache(u32 fd, u32 offset, u32 file_size)
+{
+  if (HasCacheForFile(fd, offset))
+    return 0;
+
+  u64 ticks = SimulateFlushFileCache();
+  if ((offset % CLUSTER_DATA_SIZE != 0 || offset != file_size) && offset < file_size)
+    ticks += CLUSTER_READ_TICKS;
+
+  m_cache_fd = fd;
+  m_cache_chain_index = static_cast<u16>(offset / CLUSTER_DATA_SIZE);
+  return ticks;
+}
+
+u64 FS::SimulateFlushFileCache()
+{
+  if (m_cache_fd == INVALID_FD || !m_dirty_cache)
+    return 0;
+  m_dirty_cache = false;
+  m_fd_map[m_cache_fd].superblock_flush_needed = true;
+  return CLUSTER_WRITE_TICKS;
+}
+
+// Simulate parts of the FS read/write logic to estimate ticks for file operations correctly.
+u64 FS::EstimateTicksForReadWrite(const Handle& handle, const ReadWriteRequest& request)
+{
+  u64 ticks = 0;
+
+  const bool is_write = request.command == IPC_CMD_WRITE;
+  const Result<FileStatus> status = m_ios.GetFS()->GetFileStatus(handle.fs_fd);
+  u32 offset = status->offset;
+  u32 count = request.size;
+  if (!is_write && count + offset > status->size)
+    count = status->size - offset;
+
+  while (count != 0)
+  {
+    u32 copy_length;
+    // Fast path (if not cached): FS copies an entire cluster directly from/to the request.
+    if (!HasCacheForFile(request.fd, offset) &&
+        count >= CLUSTER_DATA_SIZE && offset % CLUSTER_DATA_SIZE == 0)
+    {
+      ticks += is_write ? CLUSTER_WRITE_TICKS : CLUSTER_READ_TICKS;
+      copy_length = CLUSTER_DATA_SIZE;
+      if (is_write)
+        m_fd_map[request.fd].superblock_flush_needed = true;
+    }
+    else
+    {
+      ticks += SimulatePopulateFileCache(request.fd, offset, status->size);
+
+      const u32 start = offset - m_cache_chain_index * CLUSTER_DATA_SIZE;
+      copy_length = std::min<u32>(CLUSTER_DATA_SIZE - start, count);
+      // FS essentially just does a memcpy from/to an internal buffer.
+      ticks += 0.63f * copy_length + (request.command == IPC_CMD_WRITE ? 1000 : 0);
+      m_dirty_cache = is_write;
+
+      if (is_write && (offset + copy_length) % CLUSTER_DATA_SIZE == 0)
```

Comment:
This math makes me nervous, but I assume this case is covered in the hardware test? (i.e. that a write gets aligned to the cluster size, continues with whole clusters, and then performs the final part)

Would be nice if Dolphin had a "assert this is reached in some test case", but it would also be nice if Dolphin had actual automated tests for emulation.